### PR TITLE
Support for synchronization, generic instance running initiation/termination, retry mechanism

### PR DESCRIPTION
**Pull request by NeonKirill**
Added support for synchronization signals in MQ Connector and generic run/stop methods. Also added generic retry() decorator into utils, but did not add direct support for it, as it is noticed to cause some thread-licking issues

### CODE REVIEW

File: neon_mq_connector/utils/connection_utils.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        :param num_retries: num of retries for function execution
+        :param callback_on_exceeded: function to call when all attempts fail
+        :param callback_on_exceeded_args: args for :param callback_on_exceeded
+        :param callback_on_attempt_failure: function to call when single attempt fails
+        :param callback_on_attempt_failure_args: args for :param callback_on_attempt_failure
+        :param backoff_factor: value of backoff factor for setting delay between function execution retry,
+                               refer to "get_timeout()" for details
+    """
+
```
Comment (NeonKirill) [inline neon_mq_connector/utils/connection_utils.py:53]: todo

File: neon_mq_connector/connector.py [hunk 2]
```diff
@@ -214,3 +222,53 @@ def stop_consumers(self, names: tuple = ()):
                     self.consumers[name].join()
             except Exception as e:
                 raise ChildProcessError(e)
+
+    def sync(self, vhost: str = None, exchange: str = None, queue: str = None, request_data: dict = None):
+        """
+            Periodical notification message to be sent into MQ
```
Comment (NeonDaniel) [inline neon_mq_connector/connector.py:228]: Is there a default listener for this message? I think the docstring here should describe more what this is for.

i.e.
```
Periodically sends an MQ message to notify other services that this service is active.
```

File: neon_mq_connector/utils/connection_utils.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# China Patent: CN102017585  -  Europe Patent: EU2156652  -  Patents Pending
+import time
+
+from neon_utils import LOG
+
+
+def get_timeout(backoff_factor: float, number_of_retries: int) -> float:
+    """
+        Gets timeout based on backoff_factor
+
```
Comment (NeonDaniel) [inline neon_mq_connector/utils/connection_utils.py:27]: Add docstrings for parameters and return

File: neon_mq_connector/utils/connection_utils.py (hunk 1 above)
Comment (NeonDaniel) [reply neon_mq_connector/utils/connection_utils.py:53]: Comment known issue with this method

File: neon_mq_connector/utils/__init__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# For commercial licensing, distribution of derivative works or redistribution please contact licenses@neon.ai
+# Distributed on an "AS IS” basis without warranties or conditions of any kind, either express or implied.
+# Trademarks of Neongecko: Neon AI(TM), Neon Assist (TM), Neon Communicator(TM), Klat(TM)
+# Authors: Guy Daniels, Daniel McKnight, Elon Gasper, Richard Leeds, Kirill Hrymailo
+#
+# Specialized conversational reconveyance options from Conversation Processing Intelligence Corp.
+# US Patents 2008-2021: US7424516, US20140161250, US20140177813, US8638908, US8068604, US8553852, US10530923, US10530924
+# China Patent: CN102017585  -  Europe Patent: EU2156652  -  Patents Pending
+
+from .thread_utils import RepeatingTimer
```
Comment (NeonDaniel) [inline neon_mq_connector/utils/__init__.py:20]: Add some unit tests for these utils (or empty test classes with #TODO)

**Review by NeonDaniel — CHANGES_REQUESTED** [review]

**Comment by NeonKirill** [comment]
@NeonDaniel added testing flows and requested docstring and methods, please check if that is all fine

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        """
+            Simple method that is passing check only after n-th attempt
+            :param num_attempts: number of attempts before passing
+        """
+        if self.counter < num_attempts-1:
+            self.repeating_method()
+            raise AssertionError('Awaiting counter equal to 3')
+        return True
+
+    def test_01_get_timeout(self):
```
Comment (NeonDaniel) [inline tests/test_utils.py:56]: Do these 3 tests need to run synchronously, or can the numeric indices be removed from method names?
